// firebase-tools/lib/deploy/functions/checkRuntimeDependencies.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkRuntimeDependencies = exports.checkForNode8 = void 0;
const cli_color_1 = require("cli-color");
const track = require("../../track");
const logger = require("../../logger");
const ensureApiEnabled_1 = require("../../ensureApiEnabled");
const utils_1 = require("../../utils");
const error_1 = require("../../error");
const FAQ_URL = "https://firebase.google.com/support/faq#functions-runtime";
const CLOUD_BUILD_API = "cloudbuild.googleapis.com";
function node8DeprecationWarning() {
    track("functions_runtime_notices", "nodejs8_deprecation_warning");
    logger.warn();
    utils_1.logLabeledWarning("functions", `${cli_color_1.bold(`${cli_color_1.yellow("Warning:")} Node.js 8 functions are deprecated and will stop running on 2021-03-15.`)} Please upgrade to Node.js 10 or greater by adding an entry like this to your package.json:
    
    {
      "engines": {
        "node": "12"
      }
    }

The Firebase CLI will stop deploying Node.js 8 functions in new versions beginning ${cli_color_1.bold("2020-12-15")}, and deploys from all CLI versions will halt on ${cli_color_1.bold("2021-02-15")}. For additional information, see: ${FAQ_URL}`);
    logger.warn();
}
function nodeBillingError(projectId) {
    track("functions_runtime_notices", "nodejs10_billing_error");
    return new error_1.FirebaseError(`Cloud Functions deployment requires the pay-as-you-go (Blaze) billing plan. To upgrade your project, visit the following URL:
      
https://console.firebase.google.com/project/${projectId}/usage/details

For additional information about this requirement, see Firebase FAQs:

${FAQ_URL}`, { exit: 1 });
}
function nodePermissionError(projectId) {
    track("functions_runtime_notices", "nodejs10_permission_error");
    return new error_1.FirebaseError(`Cloud Functions deployment requires the Cloud Build API to be enabled. The current credentials do not have permission to enable APIs for project ${cli_color_1.bold(projectId)}.

Please ask a project owner to visit the following URL to enable Cloud Build:

https://console.cloud.google.com/apis/library/cloudbuild.googleapis.com?project=${projectId}

For additional information about this requirement, see Firebase FAQs:
${FAQ_URL}
`);
}
function isPermissionError(e) {
    var _a, _b, _c;
    return ((_c = (_b = (_a = e.context) === null || _a === void 0 ? void 0 : _a.body) === null || _b === void 0 ? void 0 : _b.error) === null || _c === void 0 ? void 0 : _c.status) === "PERMISSION_DENIED";
}
function checkForNode8(runtime) {
    if (runtime === "nodejs8") {
        node8DeprecationWarning();
        return;
    }
}
exports.checkForNode8 = checkForNode8;
function checkRuntimeDependencies(projectId, runtime) {
    return __awaiter(this, void 0, void 0, function* () {
        if (runtime !== "nodejs8") {
            try {
                yield ensureApiEnabled_1.ensure(projectId, CLOUD_BUILD_API, "functions");
            }
            catch (e) {
                if (error_1.isBillingError(e)) {
                    throw nodeBillingError(projectId);
                }
                else if (isPermissionError(e)) {
                    throw nodePermissionError(projectId);
                }
                throw e;
            }
        }
    });
}
exports.checkRuntimeDependencies = checkRuntimeDependencies;
